File: autozig-math/zig/affine2.zig
const std = @import("std");

pub const Affine2 = extern struct {
    matrix2: Mat2,
    translation: Vec2,

    pub const IDENTITY = Affine2{
        .matrix2 = Mat2.IDENTITY,
        .translation = Vec2{ .x = 0.0, .y = 0.0 },
    };

    pub fn from_mat2_translation(matrix2: Mat2, translation: Vec2) Affine2 {
        return .{
            .matrix2 = matrix2,
            .translation = translation,
        };
    }

    pub fn from_translation(translation: Vec2) Affine2 {
        return .{
            .matrix2 = Mat2.IDENTITY,
            .translation = translation,
        };
    }

    pub fn from_rotation(angle: f32) Affine2 {
        return .{
            .matrix2 = Mat2.from_angle(angle),
            .translation = Vec2{ .x = 0.0, .y = 0.0 },
        };
    }

    pub fn from_scale(scale: Vec2) Affine2 {
        return .{
            .matrix2 = Mat2.from_scale(scale),
            .translation = Vec2{ .x = 0.0, .y = 0.0 },
        };
    }

    pub fn from_scale_angle_translation(scale: Vec2, angle: f32, translation: Vec2) Affine2 {
        const rot = Mat2.from_angle(angle);
        const scaled = Mat2{ .cols = .{
            .{ rot.cols[0][0] * scale.x, rot.cols[0][1] * scale.x },
            .{ rot.cols[1][0] * scale.y, rot.cols[1][1] * scale.y },
        } };
        return .{
            .matrix2 = scaled,
            .translation = translation,
        };
    }

    pub fn transform_point(self: Affine2, point: Vec2) Vec2 {
        const rotated = self.matrix2.mul_vec2(point);
        return rotated.add(self.translation);
    }

    pub fn transform_vector(self: Affine2, vec: Vec2) Vec2 {
        return self.matrix2.mul_vec2(vec);
    }
};

export fn affine2_identity() Affine2 {
    return Affine2.IDENTITY;
}

export fn affine2_from_translation(translation: Vec2) Affine2 {
    return Affine2.from_translation(translation);
}

export fn affine2_from_rotation(angle: f32) Affine2 {
    return Affine2.from_rotation(angle);
}

export fn affine2_from_scale(scale: Vec2) Affine2 {
    return Affine2.from_scale(scale);
}

export fn affine2_from_scale_angle_translation(scale: Vec2, angle: f32, translation: Vec2) Affine2 {
    return Affine2.from_scale_angle_translation(scale, angle, translation);
}

export fn affine2_transform_point(self: Affine2, point: Vec2) Vec2 {
    return self.transform_point(point);
}

export fn affine2_transform_vector(self: Affine2, vec: Vec2) Vec2 {
    return self.transform_vector(vec);
}

File: autozig-math/src/affine2.rs
use autozig::include_zig;
use crate::{Vec2, Mat2};

#[repr(C)]
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Affine2 {
    pub matrix2: Mat2,
    pub translation: Vec2,
}

include_zig!("zig/affine2.zig", {
    fn affine2_identity() -> Affine2;
    fn affine2_from_translation(translation: Vec2) -> Affine2;
    fn affine2_from_rotation(angle: f32) -> Affine2;
    fn affine2_from_scale(scale: Vec2) -> Affine2;
    fn affine2_from_scale_angle_translation(scale: Vec2, angle: f32, translation: Vec2) -> Affine2;
    fn affine2_transform_point(self_: Affine2, point: Vec2) -> Vec2;
    fn affine2_transform_vector(self_: Affine2, vec: Vec2) -> Vec2;
});

impl Affine2 {
    pub const IDENTITY: Self = Self {
        matrix2: Mat2::IDENTITY,
        translation: Vec2::ZERO,
    };

    pub fn identity() -> Self {
        affine2_identity()
    }

    pub fn from_translation(translation: Vec2) -> Self {
        affine2_from_translation(translation)
    }

    pub fn from_rotation(angle: f32) -> Self {
        affine2_from_rotation(angle)
    }

    pub fn from_scale(scale: Vec2) -> Self {
        affine2_from_scale(scale)
    }

    pub fn from_scale_angle_translation(scale: Vec2, angle: f32, translation: Vec2) -> Self {
        affine2_from_scale_angle_translation(scale, angle, translation)
    }

    pub fn transform_point(self, point: Vec2) -> Vec2 {
        affine2_transform_point(self, point)
    }

    pub fn transform_vector(self, vec: Vec2) -> Vec2 {
        affine2_transform_vector(self, vec)
    }
}
